facet-reflect/src/scalar.rs

```diff
@@ -0,0 +1,217 @@
+use core::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+
+use facet_core::{ConstTypeId, Shape};
+
+/// All scalar types supported out of the box by peek and poke.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[non_exhaustive]
+pub enum ScalarType {
+    /// Unit tuple `()`.
+    Unit,
+    /// Primitive type `bool`.
+    Bool,
+    /// Primitive type `str`.
+    Str,
+    /// `std::string::String`.
+    #[cfg(feature = "std")]
+    String,
+    /// `std::borrow::Cow<'_, str>`.
+    #[cfg(feature = "std")]
+    CowStr,
+    /// Primitive type `f32`.
+    F32,
+    /// Primitive type `f64`.
+    F64,
+    /// Primitive type `u8`.
+    U8,
+    /// Primitive type `u16`.
+    U16,
+    /// Primitive type `u32`.
+    U32,
+    /// Primitive type `u64`.
+    U64,
+    /// Primitive type `usize`.
+    USize,
+    /// Primitive type `i8`.
+    I8,
+    /// Primitive type `i16`.
+    I16,
+    /// Primitive type `i32`.
+    I32,
+    /// Primitive type `i64`.
+    I64,
+    /// Primitive type `isize`.
+    ISize,
+    /// `core::net::SocketAddr`.
+    #[cfg(feature = "std")]
+    SocketAddr,
+    /// `core::net::IpAddr`.
+    IpAddr,
+    /// `core::net::Ipv4Addr`.
+    Ipv4Addr,
+    /// `core::net::Ipv6Addr`.
+    Ipv6Addr,
+    /// `facet_core::typeid::ConstTypeId`.
+    ConstTypeId,
+}
+
+impl ScalarType {
+    /// Infer the type from a shape definition.
+    pub fn try_from_shape(shape: &'static Shape) -> Option<Self> {
```

Comment:
Should this be an `Option<Self>` or panic if it's not a scalar?

If it should panic, it would also mean that it's practically exhaustive.

Reply:
Option is the right call!